drivers/network/netio/netio.c

```diff
@@ -0,0 +1,903 @@
+    PTRANSPORT_ADDRESS ta = TdiTransportAddressFromSocketAddress(LocalAddress);
+    if (ta == NULL)
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+    if (s->LocalAddressHandle != INVALID_HANDLE_VALUE)
+    {
+        ZwClose(s->LocalAddressHandle);
+        s->LocalAddressHandle = INVALID_HANDLE_VALUE;
+        s->LocalAddressFile = NULL;
+    }
+
+    status = TdiOpenAddressFile(&s->TdiName,
+                                ta, AFD_SHARE_REUSE, &s->LocalAddressHandle, &s->LocalAddressFile);
+    if (NT_SUCCESS(status))
+    {
+        memcpy(&s->LocalAddress, LocalAddress, sizeof(s->LocalAddress));
+    }
+    ExFreePoolWithTag(ta, 'ADDR');
+
+    Irp->IoStatus.Status = status;
+    return status;
+}
+
+enum direction
+{
+    DIR_SEND,
+    DIR_RECEIVE
+};
+
+static NTSTATUS WSKAPI
+WskSendTo(
+    _In_ PWSK_SOCKET Socket,
+    _In_ PWSK_BUF Buffer,
+    _Reserved_ ULONG Flags,
+    _In_opt_ PSOCKADDR RemoteAddress,
+    _In_ ULONG ControlInfoLength,
+    _In_reads_bytes_opt_(ControlInfoLength) PCMSGHDR ControlInfo,
+    _Inout_ PIRP Irp)
+{
+    PIRP tdiIrp = NULL;
+    struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL *)Socket;
+    PTDI_CONNECTION_INFORMATION TargetConnectionInfo;
+    NTSTATUS status;
+    void *BufferData;
+    struct NetioContext *nc;
+
+    if (DummyDeviceObject == NULL)
+    {
+        DbgPrint("DummyDeviceObject is NULL, was the DriverEntry funtion called?\n");
+        Irp->IoStatus.Status = STATUS_INVALID_PARAMETER;
+        return STATUS_INVALID_PARAMETER;        /* TODO: something more meaningful */
+    }
+
+    /* Call ourselves. Sets status to pending. The interesting
+     * part happens later.
+     */
+
+    DummyCallDriver(Irp);
+
+    nc = ExAllocatePoolWithTag(NonPagedPool, sizeof(*nc), 'NEIO');
+    if (nc == NULL)
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+    nc->socket = s;
+    nc->UserIrp = Irp;
+
+    TargetConnectionInfo = TdiConnectionInfoFromSocketAddress(RemoteAddress);
+    if (TargetConnectionInfo == NULL)
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    BufferData = MmGetSystemAddressForMdlSafe(Buffer->Mdl, NormalPagePriority);
+    if (BufferData == NULL)
+    {
+        DbgPrint("Error mapping MDL\n");
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    /* Some drivers (like WinDRBD) do not set this.
+     * But ReactOS seems to require this (else BSOD on completion).
+     */
+
+    if (Irp->Tail.Overlay.Thread == NULL)
+    {
+        Irp->Tail.Overlay.Thread = PsGetCurrentThread();
+    }
+    IoMarkIrpPending(Irp);
+    SocketGet(s);
+    /* TODO: protect by spinlock */
+    /* TODO: still needed? */
```

Comment:
May or may not be needed -- but definitely good for debugging. You should assert that the list is empty in `SocketPut` to make sure we're not freeing a socket with pending operations.

Reply:
Hi the PendingIrp list has been gone.